### ISSUE TITLE
FlexibleSpaceBar - no fade when scrolling

### ISSUE THREAD

**Issue by brxxxx** [flutter/flutter#128636]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

- Run my sample code on either Android or iPhone. 
- Scroll the content.
- Observe how the FlutterLogo does not fade in/out whilst scrolling.

### Expected results

According to the documentation and many examples on the internet (plus my own memory of how this previously worked) , the widget that is placed in the `background` parameter of the `FlexibleSpaceBar` should fade out as the content is scrolled up. When scrolling back down, the widget should fade in again.

### Actual results

The FlutterLogo jumps from completely visible to completely hidden (no fade).

### Code sample

<details><summary>Code sample</summary>

```import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: CustomScrollView(
        slivers: [
          const SliverAppBar(
            backgroundColor: Colors.red,
            expandedHeight: 300.0,
            pinned: true,
            flexibleSpace: FlexibleSpaceBar(
              background: FlutterLogo(),
              centerTitle: true,
            ),
          ),
          SliverList(
            delegate: SliverChildBuilderDelegate(
              childCount: 100,
              (context, index) => const SizedBox(
                height: 100,
                child: Center(
                  child: Text('something here'),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

![flexible_space_bar_not_fading_AdobeExpress](https://github.com/flutter/flutter/assets/66548748/30cec369-9455-4001-9fce-ff306037c40a)

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.10.4, on macOS 13.2.1 22D68 darwin-arm64, locale
    en-GB)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
[✗] Chrome - develop for the web (Cannot find Chrome executable at
    /Applications/Google Chrome.app/Contents/MacOS/Google Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.
[!] Android Studio (not installed)
[✓] IntelliJ IDEA Community Edition (version 2023.1.2)
[✓] Connected device (2 available)
[✓] Network resources

! Doctor found issues in 2 categories.
```

</details>

**Comment by brxxxx** [comment]
The strange thing is that I just tested the above code on DarptPad website and the animation works:

https://dartpad.dev/?id=31f558c9d6be34cd156a6ad72e136155

**Comment by lukehutch** [comment]
Same issue here. I tried every option I can find, and couldn't get the fade behavior. However, I was never trying to test on Flutter 2.x, and [this stackoverflow question](https://stackoverflow.com/questions/76115338/how-to-fix-flutters-flexiblespacebar-widget-so-it-fades-into-the-appbar-with-ma) seems to indicate that the behavior broke with Flutter 3.x.

**Comment by dam-ease** [comment]
Thanks for filing this issue @brxxxx. This seems to have been fixed on the latest `master` channel. Kindly confirm this by switching to the latest flutter `master` channel.

<details><summary>Video</summary>
<p>


https://github.com/flutter/flutter/assets/53122008/83187b32-cd2c-413e-afc7-9f38e5d0071f



</p>
</details>

**Comment by tamirrab** [comment]
On version:
Flutter 3.12.0-4.0.pre.60 • channel master
And still not fixed

```
 Scaffold(
      body: CustomScrollView(
        slivers: [
          SliverAppBar(
            pinned: true,
            leading: GestureDetector(
              onTap: () => {},
              child: const Icon(
                Icons.arrow_back,
                color: Colors.black,
              ),
            ),
            flexibleSpace: LayoutBuilder(builder: (context, constraints) {
              return FlexibleSpaceBar(
                background: Image.network(
                  "https://source.unsplash.com/random/400x225?sig=1&licors",
                  fit: BoxFit.cover,
                ),
              );
            }),
            expandedHeight: MediaQuery.of(context).size.height * 0.2,
            backgroundColor: const Color(0xFFFFF4F4),
          ),
          const SliverToBoxAdapter(child: SizedBox(height: 1000))
        ],
      ),
    );
```

**Comment by brxxxx** [comment]
> Thanks for filing this issue @brxxxx. This seems to have been fixed on the latest `master` channel. Kindly confirm this by switching to the latest flutter `master` channel.
> 
> Video
>  Simulator.Screen.Recording.-.iPhone.14.-.2023-06-12.at.09.14.31.mp4

Switched to master channel, still no fade

**Comment by brxxxx** [comment]
@dam-ease this is not fixed please see the comments

**Comment by dam-ease** [comment]
This issue seemed to have been fixed and worked very fine on master channel `3.11.0-2.0.pre.55` but no longer works on the latest master channel `3.12.0-4.0.pre.114`.

<details><summary>flutter doctor -v</summary>
<p>

```
[✓] Flutter (Channel master, 3.12.0-4.0.pre.114, on macOS 13.4 22F66 darwin-arm64, locale en-NG)
    • Flutter version 3.12.0-4.0.pre.114 on channel master at /Users/dammya/fvm/versions/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 323c296922 (58 minutes ago), 2023-06-19 03:24:11 -0400
    • Engine revision 280491d4cc
    • Dart version 3.1.0 (build 3.1.0-227.0.dev)
    • DevTools version 2.24.0

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/dammya/Library/Android/sdk
    • Platform android-33, build-tools 32.1.0-rc1
    • ANDROID_HOME = /Users/dammya/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2022.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.79.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (4 available)
    • iPhone 14 (mobile)     • 5FE8DE93-0F4C-45BA-8BED-BC9CD69B35D1 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • iPhone 14 Pro (mobile) • DCF2031A-7EA9-42FF-8CD4-D4584F3E7A6D • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 13.4 22F66 darwin-arm64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 114.0.5735.133

[✓] Network resources
    • All expected network resources are available.

• No issues found!

``` 

```
[✓] Flutter (Channel master, 3.11.0-2.0.pre.55, on macOS 13.4 22F66 darwin-arm64, locale en-NG)
    • Flutter version 3.11.0-2.0.pre.55 on channel master at /Users/dammya/fvm/versions/835b892d7f
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 835b892d7f (6 weeks ago), 2023-05-07 01:30:22 +0530
    • Engine revision 23f730efbf
    • Dart version 3.1.0 (build 3.1.0-83.0.dev)
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/dammya/Library/Android/sdk
    • Platform android-33, build-tools 32.1.0-rc1
    • ANDROID_HOME = /Users/dammya/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2022.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.79.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (4 available)
    • iPhone 14 (mobile)     • 5FE8DE93-0F4C-45BA-8BED-BC9CD69B35D1 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • iPhone 14 Pro (mobile) • DCF2031A-7EA9-42FF-8CD4-D4584F3E7A6D • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 13.4 22F66 darwin-arm64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 114.0.5735.133

[✓] Network resources
    • All expected network resources are available.

• No issues found!
``` 

</p>
</details>

**Comment by Piinks** [comment]
This looks like another regression from https://github.com/flutter/flutter/pull/116788 that was fixed by https://github.com/flutter/flutter/pull/128138 but then regressed again in https://github.com/flutter/flutter/pull/128273, which were both trying to resolve this other regression issue: https://github.com/flutter/flutter/issues/127836. That issue is ultimately fixed, this is a separate issue.

It looks like RenderOpacity is not being marked for paint so it does not repaint in response to the opacity value changing, since isRepaintBoundary is false. markNeedsCompositedLayerUpdate only ends up calling markNeedsPaint when it is and was previously a repaint boundary.

I am wondering since some many issues have stemmed from https://github.com/flutter/flutter/pull/116788 if we should revert it. @jonahwilliams what do you think? Alternatively we can add back a call to marksNeedPaint in the setter for opacity, which in some cases would be redundant, but it currently does not repaint when it should anymore.

**Comment by jonahwilliams** [comment]
Generally speaking (Woah, woah I know). There should be no problems inserting repaint boundaries in a widget tree, provided that all widgets are correcting calling set state.

I'm not certain what is happening with the Flexible space bar yet. But given that the only issue we've seen is with the flexible space bar I'm more inclined to believe there is a bug in its the implementation, rather than in the repaint boundary change. I'll take a look at this later in the week and report back.

Its possible there is a bad pattern in there that we need to understand, or it could be that the repaint boundary change overlooked something. I'd like to understand why before reverting.

**Comment by Piinks** [comment]
Thanks @jonahwilliams! SGTM, I really appreciate it. :) Happy to hold off until we better understand all the things.

**Comment by Lesiuk** [comment]
We certainly need to create and include rigorous regression tests to ensure that these issues don't creep back in.

**Comment by timmyrosen** [comment]
Any updates on this?

**Comment by Rest00rer** [comment]
Same problem. Waiting for fixes

**Comment by lukehutch** [comment]
It's not just fading that's not working -- none of the `stretchModes` (`blurBackground`, `fadeTitle`, and `zoomBackground`) are working currently.

**Comment by lukehutch** [comment]
Additionally, the background color fading seems to get stuck at something like 90% of the max brightness until the UI is "stretched" by pulling downwards at the top of the scroll zone. Note the slightly darker shade of yellow in the `SliverAppBar` compared to the full-brightness `Color.amber` of the `SliverPersistentHeader` below it, until the sliver list below these widgets is dragged all the way to the top.

https://github.com/flutter/flutter/assets/811305/ceef2885-bcd0-4c67-b069-c1ee19ac2d3b

**Comment by gabrielgarciagava** [comment]
I wonder if this is going to be cherry picked to the 3.10.7 version. 
We are currently stuck on version 3.10.3 because of such issue, while there are some dart fixes released on flutter 3.10.6 that we really would like to have.
Thank you in advance :)

**Comment by EDeLeon9** [comment]
I got affected by this issue suddenly (3.10.5 ver). I will continue developing hoping there will be a fix soon

**Comment by lukehutch** [comment]
> I got affected by this issue suddenly (3.10.5 ver). I will continue developing hoping there will be a fix soon

I had to simply duplicate the `FlexibleSpaceBar` class (it's just a regular widget) and modify it to have the behavior I needed for my specific usecase. The calculation of how much of the bar is showing is quite messed up, when you look at it closely. 

The fix for me was to make the widget animations depend only upon `constraints.maxHeight` (the currently-shown height of the widget) and `expandedHeight` from the parent `sliverAppBar`, and to throw away any reference to `settings.currentExtent` and `settings.minExtent` (which is what is messing up the animations currently).

I would post my code here, but I also stripped out every animation, pinning, and floating mode that I wasn't using, to simplify the code, so it may be a bit hard to map this back onto the original `FlexibleSpaceBar`, in order to produce a fix for the current issue.
